### DISCUSSION TITLE
How to create swagger documentation?

### DISCUSSION THREAD

**Discussion by DeanKamali** [directus/directus#7084] [Archive]
Hi everybody, 

Is there is a documented way to generate a Swagger interface for the APIs created by directus, 

I do know that strapi cms dose it, wondering if I could do the same with directus?

**Comment by benhaynes** [comment]
I may be wrong, but doesn't Swagger use the Open API Spec? If so, you can use this:

https://docs.directus.io/reference/api/system/server/#get-openapi-specification

**Reply by DeanKamali** [reply]
Are you guys considering adding api documentation to your product? 

Yes, swagger use open api standard, however looking at my project files, I have no clue where to start. 

Swagger requires to import all routes in a given project to generate documentation, wonder if there is a way to get that done

**Reply by rijkvanzanten** [reply]
We have https://docs.directus.io/reference/api/introduction/ that goes over every available endpoint in detail 👍🏻 Next to that, we do plan on utilizing the open api spec to render a dynamic API reference specific to your project under the docs module in the future

**Reply by philleepflorence** [reply]
Tangent Question: [Postman vs Swagger UI vs Insomnia](https://stackshare.io/stackups/postman-vs-swagger-ui)

What do you use or recommend for Rest API Documentation, Guides, Snippets, and Collaboration?
Might have an implementation for Postman + Directus + GitHub!
Curious to learn about Swagger UI though.

**Reply by keesvanbemmel** [reply]
[Spoiler](https://www.youtube.com/watch?v=AicEmIeuuLw&t=1055s) on this feature was made in youtube video 👍 

"... and also with an API reference that is dynamic to your schema"

**Reply by rijkvanzanten** [reply]
😏

**Reply by biapar** [reply]
> https://docs.directus.io/reference/api/introduction/

Link not valid.

**Reply by benhaynes** [reply]
https://docs.directus.io/reference/introduction/

**Comment by Guiqft** [comment]
I'm able to create a Swagger UI getting the `json` from the `/server/specs/oas` and using it to configure a new Directus endpoint like that: 

```typescript
import swaggerUi from 'swagger-ui-express';
import { EndpointRegisterFunction } from './types';

import * as swaggerDocument from './swagger.json';

const registerEndpoint: EndpointRegisterFunction = (router) => {
	router.use('/', swaggerUi.serve);
	router.get('/', swaggerUi.setup(swaggerDocument));
};

export default registerEndpoint;

```

So, when I access `${my_app_url}/custom/api-doc` I can interact with the Directus auto generated api documentation.

**Reply by arladmin** [reply]
Hi @Guiqft, firstly thanks for sharing your code.

But this isn't the code from your extensions js file, is it?
Because when i tried this (in my docker based setup, it doesn't work in my setup at all (error: 'can't use import statement', can't import the npm modules with require either).

**Reply by Guiqft** [reply]
Hi @arladmin. Yes, it's my .ts from extensions file. You should try the new `create extension` command from the directus extensions sdk and paste my code into the main .ts file. This hould work as any custom endpoint.

**Reply by keesvanbemmel** [reply]
I don't understand something. You seem to be referring to a static JSON file, however the /server/specs/oas endpoint is dynamic.

Any idea how to reference the dynamic content and pass the authenticated/authorized session to that endpoint so it'll return the JSON the user is actually to see?

The endpoint obviously doesn't have the same session as the app. I'm thinking there is no way/extension to overcome this at the moment?

FWIW, this is my extension code so far:
```
const swaggerUi = require('swagger-ui-express');

module.exports = function registerEndpoint(router, context) {
  const { ForbiddenException } = context.exceptions;
  router.use((req, res, next) => {
    if (!req.accountability.user) {
      throw new ForbiddenException();
    } else {
      next();
    }
  });

  const options = {
    swaggerOptions: {
      url: '/server/specs/oas',
    },
  };

  router.use('/', swaggerUi.serve);
  router.get('/', swaggerUi.setup(null, options));
};
```

**Reply by azrikahar** [reply]
@keesvanbemmel if we check the existing endpoint which you mentioned as dynamic, we can see it is using `SpecificationService`:

   https://github.com/directus/directus/blob/969621249fbcffe46d2106c44ad53c05a07d8d2c/api/src/controllers/server.ts#L10-L22

which is available in `context.services`, and the accountability usage here should be what you need.

Then you should be able to pass the generated Open API Object ([Reference for source code](https://github.com/directus/directus/blob/969621249fbcffe46d2106c44ad53c05a07d8d2c/api/src/services/specifications.ts#L91-L121)) from the generate() function (line 18 of the code shown above) to the `swaggerUi.setup` function.

_Disclaimer: never personally tried the above code/suggestion, but on paper it should work. Let us know how it goes_

**Comment by Skywrath-Mage** [comment]
Hello guys, 

any update on the api/endpoint documentation automatically generated by Directus? 
Now I have to use the /specs/oas endpoint and copy and past it to Swagger editor, and then save the client rendered version of the documentation. 
It's feasible, but a bit tricky. 

Have nice holidays anyway 
Matt.

**Reply by rijkvanzanten** [reply]
Hi Matt!

The API currently doesn't render out an HTML version of the spec. There's plans to include this as part of the documentation module in the app, but the timeline on that is TBD.

Hope you have a great holiday season too 🙂

**Reply by CoenWarmer** [reply]
Any updates on this? Would love to evaluate this!

**Reply by benhaynes** [reply]
I believe we already have this:
https://docs.directus.io/reference/system/server/#get-openapi-specification

**Reply by keesvanbemmel** [reply]
I think this is about HTML API docs @benhaynes

**Reply by CoenWarmer** [reply]
Hey @benhaynes, that's already great; I was hoping for a way for Directus to autogenerate a OpenAPI / Swagger HTML documentation from that spec.

**Reply by benhaynes** [reply]
Hmm, not sure about that... maybe @rijkvanzanten knows. All I know is that we are working on getting these to generate in the App (markdown -> HTML), if that's what you're looking for.

**Comment by g-30** [comment]
is there any news on this? Having an HTML page with the custom endpoints from the extensions described within the Directus panel would be incredibly helpful @benhaynes

**Reply by benhaynes** [reply]
No news... but I agree this would be a great addition to the platform!

**Reply by TBG-FR** [reply]
As stated [here](https://github.com/directus/directus/issues/2255) :
> There's currently no way to have custom endpoint inject definitions into the generated specs

I'm sure with a bit of creativity we can make it happen, and then add these new endpoints to the `/server/specs/oas` (or to a separate route ?) and reuse that data to display endpoints on the `/admin/settings/extensions` page 🙌 

I'll look into it but can't promise anything, if anyone wants to contribute, I guess all new ideas and pieces of code are welcome 👀 

Another approach would be to make a custom extension that does "only" that job, maybe

EDIT: Found that repo, will give it a try https://github.com/sacconazzo/directus-extension-api-docs
